src/librustdoc/doctest/make.rs

```diff
@@ -407,15 +407,19 @@ fn parse_source(source: &str, crate_name: &Option<&str>) -> Result<ParseSourceIn
                     push_to_s(&mut info.crate_attrs, source, attr.span, &mut prev_span_hi);
                 }
             }
+            let mut has_non_module_items = false;
             for stmt in &body.stmts {
                 let mut is_extern_crate = false;
                 match stmt.kind {
                     StmtKind::Item(ref item) => {
                         is_extern_crate = check_item(&item, &mut info, crate_name);
                     }
-                    StmtKind::Expr(ref expr) if matches!(expr.kind, ast::ExprKind::Err(_)) => {
-                        reset_error_count(&psess);
-                        return Err(());
+                    StmtKind::Expr(ref expr) => {
+                        if matches!(expr.kind, ast::ExprKind::Err(_)) {
+                            reset_error_count(&psess);
+                            return Err(());
+                        }
+                        has_non_module_items = true;
                     }
                     StmtKind::MacCall(ref mac_call) if !info.has_main_fn => {
```

Comment:
i think you need to also assume that macros can expand to expressions, not just items.

Reply:
I voluntarily ignored it for the simple fact that trying to match on a macro call is pretty much useless since we have no clue how it will be expanded. The fact that we match on `fn main` is already a pretty bad idea. If I start saying "everything that doesn't match the first condition means it's an expression", then this code would fail to compile:

```
include!("auxiliary/empty.rs");
fn main() {}
```

Because `"auxiliary/empty.rs"` would be considered an expression, meaning the whole code would be wrapped, and the compiler would be unhappy:

```
error: non-statement macro in statement position: include
  |
3 | include!("auxiliary/empty.rs");
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

So sadly, I think the best course of action here is to keep this ugly hack...

Reply:
Sure!

Reply:
Here it is: https://github.com/rust-lang/rust/issues/140310